src/extra/components/scan.py

```diff
@@ -0,0 +1,328 @@
+                if a [xarray.DataArray][] is passed it *must* have a `trainId`
+                coordinate with the train IDs (i.e. obtained with
+                [KeyData.xarray()][extra_data.KeyData.xarray]). Raw Numpy arrays are not
+                allowed.
+            resolution (float): The usable resolution of the motor. Ideally this
+                should be slightly above the noise level of the motor (and below the
+                step size of the scan). It will be guessed if not passed explicitly.
+            min_trains (int): The minimum number of trains per-step in the scan. It
+                will be guessed if not passed explicitly.
+        """
+        self._steps = []
+        self._resolution = None
+        self._min_trains = None
+
+        # Debugging variables
+        self._diff = None
+
+        from xarray import DataArray
+
+        if isinstance(motor, SourceData):
+            self._input_pos = motor["actualPosition"].xarray()
+            default_name = f"{motor.source}.actualPosition"
+        elif isinstance(motor, KeyData):
+            if not motor.ndim == 1:
+                raise ValueError(f"KeyData of motor positions must be 1-dimensional, is actually {motor.ndim}-dimensional")
+
+            self._input_pos = motor.xarray()
+            default_name = f"{motor.source}.{motor.key.removesuffix('.value')}"
+        elif isinstance(motor, DataArray):
+            if not "trainId" in motor.coords:
+                raise ValueError("DataArray of motor positions must have a trainId coordinate")
+
+            self._input_pos = motor
+            default_name = motor.name if motor.name is not None else "motor"
+        elif isinstance(motor, np.ndarray):
+            raise TypeError("The input cannot be a plain Numpy array, train ID information is required")
+
+        self._name = name if name is not None else default_name
+
+        steps = self._get_motor_steps(self._input_pos,
+                                      resolution=resolution,
+                                      min_trains=min_trains)
+
+        # Sometimes motors that have jitter are erroneously detected as a
+        # single-step scan, so we ignore those and only take scans with more
+        # than 1 step detected.
+        if len(steps) > 1:
+            self._steps = steps
+
+        self._positions = np.array([pos for pos, _ in self.steps])
+        self._positions_tids = [tids for _, tids in self.steps]
+
+    @property
+    def steps(self) -> list:
+        """List of `(position, train_ids)` tuples for each step in the scan."""
+        return self._steps
+
+    @property
+    def positions(self) -> np.ndarray:
+        """Array of positions for each step."""
+        return self._positions
+
+    @property
+    def positions_tids(self) -> list:
+        """List of train IDs for each position."""
+        return self._positions_tids
+
+    def plot(self):
```

Comment:
Could you please add an optional `axes=None` argument and skip creating a new figure if it's supplied?

Reply:
Added in 582f7d8.